### PR TITLE
Add a --dart-entrypoint-args flag to flutter run to pass through Dart entrypoint arguments on Flutter Desktop

### PR DESCRIPTION
**Pull request by gw280**
## Description

Adds a --dart-entrypoint-args multioption to flutter run.

## Related Issues

https://github.com/flutter/flutter/issues/32986

## Tests

I added the following tests:

`startApp supports dartEntrypointArgs`

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

N/A to flutter_tools

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Review by jonahwilliams — COMMENTED** [review]
I would suggest adding this to the debuggingOptions object - https://github.com/flutter/flutter/blob/d0202131fd82fdb905f5b849026ff9287a4e958b/packages/flutter_tools/lib/src/commands/run.dart#L148

The plumbing for where to send this is already complete, and it will not require any g3fixes to roll into google3

File: packages/flutter_tools/lib/src/commands/run.dart [hunk 1]
```diff
@@ -115,7 +115,12 @@ abstract class RunCommandBase extends FlutterCommand with DeviceBasedDevelopment
         help: 'Filters out all trace events except those that are specified in '
             'this comma separated list of allowed prefixes.',
         valueHelp: 'foo,bar',
-      );
+      )
+      ..addMultiOption('dart-entrypoint-args',
+        abbr: 'a',
+        help: 'Pass a list of arguments to the Dart entrypoint at application '
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/commands/run.dart:121]: I would connect this more strongly with  `main(List<String> args)`, dart entrypoint is a bit vague. I think Dart calls this the "main function".

File: packages/flutter_tools/lib/src/commands/run.dart [hunk 2]
```diff
@@ -115,7 +115,12 @@ abstract class RunCommandBase extends FlutterCommand with DeviceBasedDevelopment
...
+              'startup. Specify this option multiple times each with one argument '
+              'to pass multiple arguments to the Dart entrypoint.');
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/commands/run.dart:123]: Document that this is only support on desktop platforms (which ones).

File: packages/flutter_tools/lib/src/desktop_device.dart [hunk 3]
```diff
@@ -136,10 +136,13 @@ abstract class DesktopDevice extends Device {
       return LaunchResult.failed();
     }
 
+    List<String> command = <String>[ executable, ];
+    if (debuggingOptions?.dartEntrypointArgs != null) {
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/desktop_device.dart:140]: can `dartEntrypointArgs` be null here?

you can do this a bit more Dart-y with a spread:

```
 final Process process = await _processManager.start(
  <String>[
    executable,
    ...?debuggingOptions?.dartEntrypointArgs,
  ]
  (...)
)
```

File: packages/flutter_tools/lib/src/device.dart [hunk 4]
```diff
@@ -835,6 +835,7 @@ class DebuggingOptions {
     this.startPaused = false,
     this.disableServiceAuthCodes = false,
     this.disableDds = false,
+    this.dartEntrypointArgs,
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/device.dart:838]: A default value for each would be nice `this.dartEntrypointArgs = const <String>[]`

File: packages/flutter_tools/lib/src/commands/run.dart (hunk 1 above)
Comment (gw280) [reply packages/flutter_tools/lib/src/commands/run.dart:121]: The Dart entrypoint can be any name, the default is main. I will clarify that it's main() by default though.

Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/commands/run.dart:121]: I guess we might have made the configurable in our embedder? But Dart says it is only `main()`

**Review by jonahwilliams — APPROVED** [review]
LGTM

**Comment by blurymind** [comment]
is there any example code to show how to use this to do `myApp.exe /path/to/file.txt `?

**Comment by tvolkert** [comment]
If we have this flag in the shared runner, it would be nice if it worked for all platforms.  (I tried it for Android, and it doesn't do anything).

**Pull request merged**